crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -326,6 +401,181 @@ impl IndexeddbCryptoStore {
             serde_wasm_bindgen::from_value(stored_request)?;
         Ok(self.serializer.deserialize_value_from_bytes(&idb_object.request)?)
     }
+
+    /// Process all the changes and do all encryption/serialization before the
+    /// actual transaction.
+    async fn prepare_for_transaction(&self, changes: &Changes) -> Result<PendingIndexeddbChanges> {
+        let mut indexeddb_changes = PendingIndexeddbChanges::new();
+
+        let private_identity_pickle =
+            if let Some(i) = &changes.private_identity { Some(i.pickle().await) } else { None };
+
+        let decryption_key_pickle = &changes.backup_decryption_key;
+        let backup_version = &changes.backup_version;
+
+        let mut core = indexeddb_changes.get(keys::CORE);
+        if let Some(next_batch) = &changes.next_batch_token {
+            core.put(
+                JsValue::from_str(keys::NEXT_BATCH_TOKEN),
+                self.serializer.serialize_value(next_batch)?,
+            );
+        }
+
+        if let Some(i) = &private_identity_pickle {
+            core.put(
+                JsValue::from_str(keys::PRIVATE_IDENTITY),
+                self.serializer.serialize_value(i)?,
+            );
+        }
+
+        if let Some(a) = &decryption_key_pickle {
+            indexeddb_changes.get(keys::BACKUP_KEYS).put(
+                JsValue::from_str(keys::RECOVERY_KEY_V1),
+                self.serializer.serialize_value(&a)?,
+            );
+        }
+
+        if let Some(a) = &backup_version {
+            indexeddb_changes
+                .get(keys::BACKUP_KEYS)
+                .put(JsValue::from_str(keys::BACKUP_KEY_V1), self.serializer.serialize_value(&a)?);
+        }
+
+        if !changes.sessions.is_empty() {
+            let mut sessions = indexeddb_changes.get(keys::SESSION);
+
+            for session in &changes.sessions {
+                let sender_key = session.sender_key().to_base64();
+                let session_id = session.session_id();
+
+                let pickle = session.pickle().await;
+                let key = self.serializer.encode_key(keys::SESSION, (&sender_key, session_id));
+
+                sessions.put(key, self.serializer.serialize_value(&pickle)?);
+            }
+        }
+
+        if !changes.inbound_group_sessions.is_empty() {
+            let mut sessions = indexeddb_changes.get(keys::INBOUND_GROUP_SESSIONS_V3);
+
+            for session in &changes.inbound_group_sessions {
+                let room_id = session.room_id();
+                let session_id = session.session_id();
+                let key = self
+                    .serializer
+                    .encode_key(keys::INBOUND_GROUP_SESSIONS_V3, (room_id, session_id));
+                let value = self.serialize_inbound_group_session(session).await?;
```

Comment:
@andybalaam Just to double check with you that it's correct. I have just merged the changes that introduces the inbound store

Reply:
Yep, looks right to me :+1: